Doc/library/os.rst

```diff
@@ -2839,18 +2839,75 @@ features:
       ``follow_symlinks=False`` had been specified instead of raising an error.
 
 
+.. function:: statx(path, mask, *, dir_fd=None, follow_symlinks=True, flags=0)
+
+   Get selected fields of the status of a file or a file descriptor. Perform the
+   equivalent of a :c:func:`statx` system call on the given path. *path* may be
+   specified as either a string or bytes -- directly or indirectly through the
+   :class:`PathLike` interface -- or as an open file descriptor. Return a
+   :class:`stat_result` object.
+
+   *mask* is a bitwise combination of the ``STATX_*`` attributes in the
+   :mod:`stat` module, indicating which fields the caller intends to use. Note
+   that the set of fields returned may differ from what's requested, if the
+   operating system or file system does not support the metadata, or if it can
+   provide additional fields with no extra effort.
+
+   This function normally follows symlinks; to stat a symlink add the argument
+   ``follow_symlinks=False``.
+
+   This function can support :ref:`specifying a file descriptor <path_fd>` and
+   :ref:`not following symlinks <follow_symlinks>`.
+
+   On Windows, passing ``follow_symlinks=False`` will disable following all
+   name-surrogate reparse points, which includes symlinks and directory
+   junctions. Other types of reparse points that do not resemble links or that
+   the operating system is unable to follow will be opened directly. When
+   following a chain of multiple links, this may result in the original link
+   being returned instead of the non-link that prevented full traversal. To
+   obtain stat results for the final path in this case, use the
+   :func:`os.path.realpath` function to resolve the path name as far as
+   possible and call :func:`lstat` on the result. This does not apply to
+   dangling symlinks or junction points, which will raise the usual exceptions.
```

Comment:
I'd prefer for the `follow_symlinks` behavior to be discussed in the ["Files and Directories"](https://docs.python.org/3/library/os.html#files-and-directories) section instead of in the `stat()` and `statx()` documentation. If we add `follow_symlinks` support to other os functions (e.g. `os.access()`), the behavior had better be consistent with `os.stat()`. For example:

```text
* **not following symlinks:** If *follow_symlinks* is
  ``False``, and the last element of the path to operate on is a symbolic link,
  the function will operate on the symbolic link itself rather than the file
  pointed to by the link.  (For POSIX systems, Python will call the ``l...``
  variant of the function.)

  You can check whether or not *follow_symlinks* is supported for a particular
  function on your platform using :data:`os.supports_follow_symlinks`.
  If it's unavailable, using it will raise a :exc:`NotImplementedError`.

  On Windows, ``follow_symlinks`` applies to name-surrogate reparse points,
  including symlinks and junctions. This type of reparse point is a symbolic
  link to another path on the system.  All other types of reparse point (e.g.
  tiered storage) are always traversed, except that reparse points that are
  not supported by the system are operated on directly.

  If a path is either a symlink or a junction that cannot be traversed to the
  final path, :func:`os.path.realpath` can be used to resolve as much of the
  target path as possible.
```
This example simplifies the discussion a bit. The chain of links discussion is moot. `stat()` will fail if the path is a symlink or junction that can't be resolved, regardless of whether it chains to an unsupported reparse point or if some target device/file/directory doesn't exist or is inaccessible. Other reparse points are black boxes. If they fail for any reason, there's nothing we can do to resolve the path beyond the reparse point.